### ISSUE TITLE
Compile error in cube type impl if struct behind conditional

### ISSUE THREAD

**Issue by torsteingrindvik** [tracel-ai/cubecl#333]
I minified an example to roughly the methods `foo` and `bar` in the code below.
I'd expect them both to work since they achieve similar things but only `bar` works.

```rust
#[derive(CubeType)]
struct Foo {
    a: f32,
}

#[cube]
impl Foo {
    // mismatched types
    // expected struct `ExpandElementTyped<_>`
    //    found struct `FooExpand`
    pub fn foo(&self, v: u32) -> Foo {
        if v > 0 {
            Foo { a: 1.0 }
        } else {
            Foo { a: 0.0 }
        }
    }

    // ok
    pub fn bar(&self, v: u32) -> Foo {
        let mut foo = Foo { a: 0.0 };

        if v > 0 {
            foo.a = 1.0
        }

        foo
    }
}
```

I'm guessing this is related to the conditional in the expansion, see image:

![image](https://github.com/user-attachments/assets/9e502604-cb6e-4d05-a58f-a44f96f7198e)

**Comment by nathanielsimard** [comment]
Yeah it can't work since the condition isn't constant. I could work if `v` was tagged as `comptime`.

**Comment by torsteingrindvik** [comment]
> Yeah it can't work since the condition isn't constant. I could work if v was tagged as comptime.

Ok- it's not obvious to me why. So I'm thinking others users will inevitably end up with similar errors.

Is it feasible to emit a good span from the proc macro if this happens?
Along with a helpful error msg this corner case would be less confusing.

**Comment by nathanielsimard** [comment]
That would be an improvement! Still unsure how to make good error messages when the error comes from the generated code.
